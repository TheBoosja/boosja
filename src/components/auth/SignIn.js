import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Field, reduxForm } from 'redux-form';
import { signInUser, authError } from '../../actions/auth';
import { Form, Error, Button, Title } from '../styled';
import renderField from './renderField';

export class SignIn extends Component {
	constructor(props) {
		super(props);

		this.onFormSubmit = this.onFormSubmit.bind(this);
	}

	componentWillUnmount() {
		this.props.authError('');
	}

	renderAlert() {
		const { error } = this.props;
		return error && <Error>{error}</Error>;
	}

	onFormSubmit({ email, password }) {
		this.props.signInUser({ email, password });
	}

	render() {
		const { handleSubmit } = this.props;

		return (
			<Form onSubmit={handleSubmit(this.onFormSubmit)}>
				<Title>Sign In</Title>
				<Field
					label='Email'
					name='email'
					type='email'
					autoFocus
					component={renderField} />
				<Field
					label='Password'
					name='password'
					type='password'
					component={renderField} />
				<Button large type='submit'>
					Sign In
				</Button>
				{this.renderAlert()}
			</Form>
		);
	}
}

function mapStateToProps({ auth: { error } }) {
	return { error };
}

export default reduxForm({
	form: 'signin'
})(connect(mapStateToProps, { signInUser, authError })(SignIn));