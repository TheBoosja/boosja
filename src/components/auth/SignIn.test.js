import React from 'react';
import { shallow } from 'enzyme';
import { SignIn } from './SignIn';

describe('Sign In', () => {
	const mockSignInUser = jest.fn();
	const mockHandleSubmit = jest.fn();
	const mockAuthError = jest.fn();

	const props = { 
		signInUser: mockSignInUser, 
		handleSubmit: mockHandleSubmit,
		authError: mockAuthError
	};
	const signIn = shallow(<SignIn {...props} />);

	it('renders properly', () => {
		expect(signIn).toMatchSnapshot();
	});

	it('contains a Form element', () => {
		expect(signIn.find('Form').exists()).toBe(true);
	});

	it('contains a Field of type email', () => {
		expect(signIn.find({ type: 'email' }).exists()).toBe(true);
	});

	it('contains a Field of type password', () => {
		expect(signIn.find({ type: 'password' }).exists()).toBe(true);
	});

	it('contains a submit Button', () => {
		expect(signIn.find('Button').props().type).toEqual('submit');
	});

	describe('when the user submits the Sign In form', () => {
		it('calls the `signInUser()` with email and password', () => {
			const cred = { email: 'test@mail.com', password: '1234' };
			signIn.instance().onFormSubmit(cred);

			expect(mockSignInUser).toHaveBeenCalledWith(cred);
		});
	});

	describe('when component will unmount', () => {
		it('resets any error from server', () => {
			signIn.instance().componentWillUnmount();
			expect(mockAuthError).toHaveBeenCalledWith('');
		});
	});

	describe('when an error doesn\'t occur', () => {
		it('does not show an error', () => {
			expect(signIn.find('Error').exists()).toBe(false);
		});
	});

	describe('when an error occurs', () => {
		it('shows the error', () => {
			const error = { error: 'An error occurred. Do something!'};
			signIn.setProps(error);

			expect(signIn.find('Error').children().text()).toEqual(error.error);
		});
	});
});