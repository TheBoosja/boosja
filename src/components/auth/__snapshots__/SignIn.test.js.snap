// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sign In renders properly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SignIn
    authError={[Function]}
    handleSubmit={[Function]}
    signInUser={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <Title>
          Sign In
</Title>,
        <Field
          autoFocus={true}
          component={[Function]}
          label="Email"
          name="email"
          type="email"
/>,
        <Field
          component={[Function]}
          label="Password"
          name="password"
          type="password"
/>,
        <Button
          large={true}
          type="submit"
>
          Sign In
</Button>,
        undefined,
      ],
      "onSubmit": undefined,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Sign In",
        },
        "ref": null,
        "rendered": "Sign In",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "autoFocus": true,
          "component": [Function],
          "label": "Email",
          "name": "email",
          "type": "email",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "component": [Function],
          "label": "Password",
          "name": "password",
          "type": "password",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Sign In",
          "large": true,
          "type": "submit",
        },
        "ref": null,
        "rendered": "Sign In",
        "type": [Function],
      },
      undefined,
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <Title>
            Sign In
</Title>,
          <Field
            autoFocus={true}
            component={[Function]}
            label="Email"
            name="email"
            type="email"
/>,
          <Field
            component={[Function]}
            label="Password"
            name="password"
            type="password"
/>,
          <Button
            large={true}
            type="submit"
>
            Sign In
</Button>,
          undefined,
        ],
        "onSubmit": undefined,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Sign In",
          },
          "ref": null,
          "rendered": "Sign In",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "autoFocus": true,
            "component": [Function],
            "label": "Email",
            "name": "email",
            "type": "email",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "component": [Function],
            "label": "Password",
            "name": "password",
            "type": "password",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Sign In",
            "large": true,
            "type": "submit",
          },
          "ref": null,
          "rendered": "Sign In",
          "type": [Function],
        },
        undefined,
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
